### PR TITLE
Implementation of optional compare(expected=..., actual=...) params

### PR DESCRIPTION
**Pull request by asqui**
Providing one or both of these named arguments allows compare()
to give more meaningful error messages on failure, which disambiguate
between the expected and actual values.

Added 42 new unit tests to cover the new functionality across all
comparison types. General documentation not added yet but this would
need to be done before releasing the feature.

(Also made some of the existing multi-line tests a bit more readable by
using textwrap.dedent() in conjunction with multi-line strings so that you
can actually see what errors will actually look like in the console rather
than just a jumble of escape codes!)

### CODE REVIEW

**Comment by cjw296** [comment]
Ignore the AppVeyor failure above, but you'll need to get that one Python 2.6 test passing...

What code coverage do the 42 new tests add? I'm not keen to have them unless they cover something that wasn't covered before...

File: testfixtures/comparison.py [hunk 1]
```diff
@@ -11,12 +11,16 @@
...
+
 def compare_simple(x, y, context):
     """
     Returns a very simple textual difference between the two supplied objects.
     """
-    return '%r != %r' % (x, y)
+    return '%r%s != %s%r' % (x,
+                             context.maybe_include(' (expected)'),
+                             context.maybe_include('(actual) '),
+                             y)
```
Comment (cjw296) [inline testfixtures/comparison.py]: Hmm, not sure I like all this added complexity...

File: testfixtures/comparison.py [hunk 2]
```diff
@@ -384,7 +419,45 @@ def compare(x, y, **kw):
...
+        raise ValueError('Must specify x and y together, '
+                         'or x with one of expected/actual.')
+
+    if expected is not_there and actual is not_there:
+        return x, y, False
+    if x is not_there and y is not_there:
+        return expected, actual, True
+    if actual is not_there:
+        return expected, x, True
+    return x, actual, True
```
Comment (cjw296) [inline testfixtures/comparison.py]: DEFINITELY not a fan of this huge extra method.


I was envisaging more:

```python
def compare(*args, **kwargs):
    ...
    if args:
        x, y = args
   else:
        x = kwargs.pop('expected')
        y = kwargs.pop('actual')
```

Not sure anything more fancy is required.

Definitely not keen to support anything other than `compare(a, b, ...)` or `compare(expected=a, actual=b)`.

File: testfixtures/comparison.py [hunk 3]
```diff
@@ -11,12 +11,16 @@
     )
 from testfixtures.resolve import resolve
 from types import GeneratorType
-    
+
+
```
Comment (cjw296) [inline testfixtures/comparison.py]: No need for the whitespace changes, thanks ;-)

File: testfixtures/tests/test_compare.py [hunk 4]
```diff
@@ -29,23 +32,41 @@ def hexsub(raw):
...
+        with ShouldRaise(ValueError(msg)):
+            compare()
+        with ShouldRaise(ValueError(msg)):
+            compare(None)
+        with ShouldRaise(ValueError(msg)):
+            compare(y=None)
+        with ShouldRaise(ValueError(msg)):
+            compare(expected=None)
+        with ShouldRaise(ValueError(msg)):
+            compare(actual=None)
```
Comment (cjw296) [inline testfixtures/tests/test_compare.py]: These should be separate tests, but hopefully with a simpler approach, less will be needed.

File: testfixtures/tests/test_compare.py [hunk 5]
```diff
@@ -150,28 +312,71 @@ def test_sequence_long(self):
...
+                second:
+                ['quite a long string 9',
+                 'quite a long string 10',
+                 'quite a long string 11',
+                 'quite a long string 12']
+
+                While comparing [4]: 
+                'quite a long string 5'
+                !=
+                'quite a long string 9'"""))
```
Comment (cjw296) [inline testfixtures/tests/test_compare.py]: sorry, but I prefer the original style, please can you leave these changes out of the pull request.

**Comment by cjw296** [comment]
I think having (expected) (actual) in the text is too noisy.
If you can get it so expected is always on the left or first and actual is always on the right, or second, that will work better.

Getting there though!

**Comment by asqui** [comment]
Hi Chris,

On the comments regarding the design.

My main design goals in this change were:
1. Make the output explicitly indicate which is the expected and which is the actual values; and
2. Make it as easy as possible for existing and new code to "opt-in" to the explicit expected/actual values.

Let's talk about design point 1 first and agree on that.

My goal was that you can tell from looking at test failure output alone which way around the expected vs. actual values are. The only way to do this is to mention expected and actual in the output. If you rely on ordering, then you rely on:
1. All developers being aware of this convention -- if this was easily feasible then none of these changes are necessary, because everyone could just use a consistent ordering in all tests; or
2. Reading the code to work out if the code has opted in to the standard ordering -- this also defeats the purpose of the change because if I'm looking at the code then I can see which is passed first and which is passed second.

The whole point of the change is to be able to understand what is going on from looking at test output alone. If you don't make this explicit in the output then, in my opinion, there's no point in making any changes.

Your thoughts?

**Comment by cjw296** [comment]
So, fundamentally, if a test is failing you will need to look at both 
the test and code under test.
You're going to have to change one or other to get the test to pass...
As such, I don't see much benefit it adding text to the failure message 
to highlight which part is expected and which is actual.

There is, for me, actually a downside to having even more text shown on 
failure as it adds to the clutter and you need to figure out what's 
boilerplate and what's actually important data.

I guess I misunderstood and thought you just wanted to be able to 
`compare(expected=..., actual=...)`, for which I do see some benefit in 
being explicit.

Not sure where that leaves us. As an option, if you're passionate about 
the text appearing, you could register your own comparers where you find 
that adds most value.

Your thoughts?

**Comment by cjw296** [comment]
Different implementation released in 4.4.0

**Pull request closed without merging**